Add Teleport Discover metrics RFD

r0mant opened the pull request.

r0mant committed:
Add Teleport Discover metrics RFD

marcoandredinis commented on rfd/0094-discover-metrics.md:
```diff
@@ -0,0 +1,195 @@
+}
+
+// Contains common metadata identifying resource type being added.
+message ResourceMetadata {
+    // Resource type: server, kubernetes, database, app, desktop.
+    string type = 1;
+    // Optional protocol where applicable, e.g. postgres, mysql, sqlserver.
+    string protocol = 2;
+}
```
This can be added later on, when we add those resources but:

Should we also add cloud and cloud product name?
It might be `postgres`, but then it's running in AWS and using the Aurora or RDS

r0mant replied:
Yeah we should be able to differentiate between these deployment modes. I've made an update to use a more flat structure so instead of type+protocol+cloud we can just use a single field that would map onto our Discover wizard "option" such as "Self-hosted Postgres" or "Azure SQL Server". Then we should be able to analyze each track individually and build a funnel for, say, databases in general too.

marcoandredinis commented on rfd/0094-discover-metrics.md:
```diff
@@ -0,0 +1,195 @@
+    // Indicates the step outcome. For example, "success" means user proceeded
+    // to the next step, "error" if something failed (e.g. connection test),
+    // "aborted" if user exited the wizard.
+    string status = 1;
```
I think we should use Enum here
Should be to normalize the query values in Posthog

r0mant replied:
GRPC is just our internal communication format for these events, when converted to Posthog events these fields would become properties. But I updated to use enum anyway.

marcoandredinis commented on rfd/0094-discover-metrics.md:
```diff
@@ -0,0 +1,195 @@
+    Status status = 3;
+}
+
+// Emitted when a user deployed a database agent and proceeded to the next step.
```
We have another step here: setting the labels and generating a token (happens in the same screen)
Should we add an event for that as well?

r0mant replied:
I thought about the same but wasn't sure if that actually provides a lot of value for now. TBH I think we could've just started with "start" and "stop" event and that already would be better than what we have now, but eventually I've settled for "1 event per screen" which should give us enough granularity to understand how users flow through the wizard.

marcoandredinis commented on rfd/0094-discover-metrics.md:
```diff
@@ -0,0 +1,195 @@
+message DiscoverTestConnectionEvent {
+    DiscoverMetadata discover_metadata = 1;
+    ResourceMetadata resource_metadata = 2;
+    Status status = 3;
```
Should we add the failing traces in case of failure?
I think they would be really useful: was it a RBAC check, connectivity issue, host/db refused the login user

r0mant replied:
Yes, the idea was that this information will make it in the "error" field on the "status".

r0mant committed:
Fix

espadolini reviewed (approved):
Should we also add the anonymized user that's going through Discover, so we can also see how many different users in the cluster actually try it? Are the Discover sessions unique per user?

kimlisa commented on rfd/0094-discover-metrics.md:
```diff
@@ -0,0 +1,214 @@
+    // "aborted" if user exited the wizard.
+    DiscoverStatus status = 1;
+    // Contains error details. We have to be careful to not include any
+    // identifyable infomation like server addresses here.
```
```suggestion
    // identifiable information like server addresses here.
```

marcoandredinis committed:
Update Events

This commit includes some nit changes:
- adding the unspecified as the first enum
- renaming some messages to add the Discover namespace

Some other changes:
- added the Skipped status to accomodate for when we don't have to
  install the DatabaseService
- merged deploy of services into a single event (deploy of an ssh agent,
  kube agent and database agent)
- added the started event to capture when the wizard is launched

r0mant commented:
@marcoandredinis @ryanclark Can we merge this RFD now since we already started the implementation?

marcoandredinis committed:
Remove prefix on enums

marcoandredinis commented:
Remove the enums prefix and renamed the RFD Id to 104 (seems to be next available slot and the 94 was taken by some other rfd)
